### ISSUE TITLE
Bitbake build failure on open bmc for romulus target 

### ISSUE THREAD

**Issue by rohitpai** [openbmc/openbmc#3840]
I am trying to build the basic image for romulus target using the basic instruction in the dev document page. 
The build is failing with below compilation error. 

`ERROR: phosphor-fan-1.0+gitAUTOINC+21dc91f693-r1 do_compile: oe_runmake failed
ERROR: phosphor-fan-1.0+gitAUTOINC+21dc91f693-r1 do_compile: ExecutionError('/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/temp/run.do_compile.3157', 1, None, None)
ERROR: Logfile of failure stored in: /home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/temp/log.do_compile.3157
`
`/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/recipe-sysroot-native/usr/bin/arm-openbmc-linux-gnueabi/../../libexec/arm-openbmc-linux-gnueabi/gcc/arm-openbmc-linux-gnueabi/11.2.0/ld: /home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/recipe-sysroot/usr/lib/Scrt1.o: in function `_start':
| /usr/src/debug/glibc/2.34-r0/git/csu/../sysdeps/arm/start.S:118: undefined reference to `main'
| collect2: error: ld returned 1 exit status
| make[3]: *** [Makefile:673: fanctl] Error 1
| make[3]: *** Waiting for unfinished jobs....
| /home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/recipe-sysroot-native/usr/bin/arm-openbmc-linux-gnueabi/../../libexec/arm-openbmc-linux-gnueabi/gcc/arm-openbmc-linux-gnueabi/11.2.0/ld: /home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/recipe-sysroot/usr/lib/Scrt1.o: in function `_start':
| /usr/src/debug/glibc/2.34-r0/git/csu/../sysdeps/arm/start.S:118: undefined reference to `main'
| collect2: error: ld returned 1 exit status
| make[3]: Leaving directory '/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/build/control'
| make[3]: *** [Makefile:772: phosphor-fan-control] Error 1
| make[2]: *** [Makefile:588: all] Error 2
| make[2]: Leaving directory '/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/build/control'
| make[1]: Leaving directory '/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/build'
| make[1]: *** [Makefile:506: all-recursive] Error 1
| make: *** [Makefile:420: all] Error 2
| ERROR: oe_runmake failed
| WARNING: exit code 1 from a shell command.
ERROR: Task (/home/rohitpai/workdir/bmc/openbmc/meta-phosphor/recipes-phosphor/fans/phosphor-fan_git.bb:do_compile) failed with exit code '1'
`

Please let me know know how I can fix it. 
thanks

**Comment by msbarth** [comment]
What does it show in `/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/romulus-openbmc-linux-gnueabi/phosphor-fan/1.0+gitAUTOINC+21dc91f693-r1/temp/log.do_compile.3157` as the compile failure?

It appears to be something to do with the `fanctl` tool included with fan control getting linked but there may be more info in that log file.

**Comment by rohitpai** [comment]
Thank you Matthew for your reply. I tried cleaning the phosphor-fan recipe and built again then the same problem did not appear. I assume something was broken and clean and rebuild has fixed the issue. 

I am facing a new issue with next build step.  The component phosphor-webui is not getting built. Here is the log. 
`npm timing 
config:load:defaults Completed in 1ms
npm
 timing config:load:file:/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/all-openbmc-linux/phosphor-webui/1.0+gitAUTOINC+f59274e8ec-r1/recipe-sysroot-native/usr/lib/node_modules/npm/npmrc Completed in 2ms
npm
 timing config:load:builtin Completed in 2ms
npm WARN invalid config proxy="" set in command line options
npm
 WARN invalid config Must be full url with "http://"
npm WARN 
invalid config https-proxy="" set in command line options
npm WARN invalid config Must be full url with "http://"
npm timing config:load:cli Completed in 3ms
npm timing config:load:env Completed in 0ms

npm timing config:load:file:/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/all-openbmc-linux/phosphor-webui/1.0+gitAUTOINC+f59274e8ec-r1/git/.npmrc Completed in 1ms
npm timing config:load:project Completed in 1ms
npm timing config:load:file:/home/rohitpai/.npmrc Completed in 1ms
npm timing config:load:user Completed in 1ms
npm timing config:load:file:/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/all-openbmc-linux/phosphor-webui/1.0+gitAUTOINC+f59274e8ec-r1/recipe-sysroot-native/usr/etc/npmrc Completed in 0ms
npm timing config:load:global Completed in 0ms
npm timing config:load:validate Completed in 0ms
npm timing config:load:credentials Completed in 1ms
npm timing config:load:setEnvs Completed in 0ms
npm timing config:load Completed in 9ms
npm timing npm:load:configload Completed in 9ms
npm timing npm:load:setTitle Completed in 0ms
npm timing npm:load:setupLog Completed in 1ms
npm timing config:load:flatten Completed in 2ms
npm timing npm:load:cleanupLog Completed in 1ms

npm timing npm:load:configScope Completed in 0ms

npm timing
 npm:load:projectScope Completed in 0ms
npm timing npm:load Completed in 16ms

npm timing arborist:ctor Completed in 1ms

npm timing idealTree:init Completed in 424ms

npm WARN old lockfile 
npm WARN old lockfile
 The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 

npm WARN 
old lockfile FetchError: request to http://registry.npmjs.org/@babel%2fcode-frame failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org
npm WARN old lockfile     at ClientRequest.<anonymous> (/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/all-openbmc-linux/phosphor-webui/1.0+gitAUTOINC+f59274e8ec-r1/recipe-sysroot-native/usr/lib/node_modules/npm/node_modules/minipass-fetch/lib/index.js:110:14)
npm WARN old lockfile     at ClientRequest.emit (node:events:390:28)
npm WARN old lockfile     at Socket.socketErrorListener (node:_http_client:447:9)
npm WARN old lockfile     at Socket.emit (node:events:402:35)
npm WARN old lockfile     at emitErrorNT (node:internal/streams/destroy:157:8)
npm WARN old lockfile     at emitErrorCloseNT (node:internal/streams/destroy:122:3)
npm WARN old lockfile     at processTicksAndRejections (node:internal/process/task_queues:83:21)
npm WARN old lockfile  Could not fetch metadata for @babel/code-frame@7.8.3 FetchError: request to http://registry.npmjs.org/@babel%2fcode-frame failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org
npm WARN old lockfile     at ClientRequest.<anonymous> (/home/rohitpai/workdir/bmc/openbmc/build/romulus/tmp/work/all-openbmc-linux/phosphor-webui/1.0+gitAUTOINC+f59274e8ec-r1/recipe-sysroot-native/usr/lib/node_modules/npm/node_modules/minipass-fetch/lib/index.js:110:14)
npm WARN old lockfile     at ClientRequest.emit (node:events:390:28)
npm WARN old lockfile     at Socket.socketErrorListener (node:_http_client:447:9)
npm WARN old lockfile     at Socket.emit (node:events:402:35)
npm WARN old lockfile     at emitErrorNT (node:internal/streams/destroy:157:8)
npm WARN old lockfile     at emitErrorCloseNT (node:internal/streams/destroy:122:3)
npm WARN old lockfile     at processTicksAndRejections (node:internal/process/task_queues:83:21) {
npm WARN old lockfile   code: 'EAI_AGAIN',
npm
 WARN old lockfile   errno: 'EAI_AGAIN',
npm WARN old lockfile   syscall: 'getaddrinfo',
npm WARN old lockfile   hostname: 'registry.npmjs.org',
npm WARN old lockfile   type: 'system'
npm WARN old lockfile }
`

From the command line I am able to ping registry.npmjs.org and there is no issue with DNS or proxy. I tried all the solution given in the internet for similar kind of issue but nothing seem to work for me. 
Have you faced similar issue before ?
Does the build require any specific version of NPM ? mine is currently 8.3.1

I truely appriciate the help.

**Comment by msbarth** [comment]
I have not seen a similar issue. From a quick search it seems this would be related to your network.
From following the recipe(https://github.com/openbmc/openbmc/blob/master/meta-phosphor/recipes-phosphor/webui/phosphor-webui_git.bb) to the phosphor-webui repo(https://github.com/openbmc/phosphor-webui) it had a README denoting `npm (>= 6.0.1)` so you should be fine there.

**Comment by rohitpai** [comment]
Yes indeed its a network failure but looks like an intended one from bitbake.  There is no network failure in the host/build  PC. 
Before starting any configure and compile task of the reciep bitbake seem to disable the network internal to bitbake server execution and all network calls will fail inside these tasks. I tried normal ping with IP this also failed. 
As a hack I cached all the missing registry info inside the fetch task so that when compile gets executed the registry info would be obtained from the cache. 
The problem seems to be coming from npm identifying package-lock.json to be in old format and trying to fix it by fetching extra info during compile task which fail because network is disabled by bitbake. 
If the issue is confirmed I can work on the fix. 
Finally I am able to build the image succesfully.

**Comment by msbarth** [comment]
If you do confirm the issue and determine a fix, note that the openbmc project uses gerrit to submit changes and not thru github.
https://github.com/openbmc/docs/blob/master/CONTRIBUTING.md
https://gerrit.openbmc-project.xyz/

**Comment by rohitpai** [comment]
looks like the issue came from recent commit [https://github.com/openbmc/openbmc/commit/595f63084d313c9708cb435470e6c6dcbbbbe646](this)

changes I think shoud be as below

`diff --git a/poky/bitbake/bin/bitbake-worker b/poky/bitbake/bin/bitbake-worker
index 3aaf3c244..5188bd7d2 100755
--- a/poky/bitbake/bin/bitbake-worker
+++ b/poky/bitbake/bin/bitbake-worker
@@ -260,10 +260,10 @@ def fork_off_task(cfg, data, databuilder, workerdata, fn, task, taskname, taskha
                 the_data.setVar('BB_UNIHASH', unihash)
 
                 bb.utils.set_process_name("%s:%s" % (the_data.getVar("PN"), taskname.replace("do_", "")))
-
-                if not the_data.getVarFlag(taskname, 'network', False):
-                    logger.debug("Attempting to disable network")
-                    bb.utils.disable_network(uid, gid)
+                if the_data.getVarFlag(taskname, 'network', False) is not None:
+                    if not the_data.getVarFlag(taskname, 'network', False):
+                        logger.debug("Attempting to disable network")
+                        bb.utils.disable_network(uid, gid)
 
                 # exported_vars() returns a generator which *cannot* be passed to os.environ.update() 
                 # successfully. We also need to unset anything from the environment which shouldn't be there 
`

**Comment by williamspatrick** [comment]
> looks like the issue came from recent commit [https://github.com/openbmc/openbmc/commit/595f63084d313c9708cb435470e6c6dcbbbbe646](this)

This change is where we sync from upstream Yocto and it seems you’re suggesting a fix of some sort in bitbake itself (I can’t actually tell from the code snippet you’ve quoted). I suspect that is highly unlikely to be the issue, but if it really needs to be done then we need to submit it upstream to Yocto.

**Comment by rohitpai** [comment]
Yes the change in bitbake-worker seems to be pulled from upstream and I will check if I can push for a patch there. 
To get my phoshphor-webui working I had to below change. 

`python __anonymous() {
  d.appendVarFlag('do_compile', 'network', '1') 
}

do_compile () {
    cd ${S}
    rm -rf node_modules
    npm --loglevel info --proxy=${http_proxy} --https-proxy=${https_proxy} install
    npm run-script build
}`

I am not an expert in Yocto but what I found was when the network flag is not set and compile executed the network access is blocked and the compilation will not work. Either we need a fix in up-stream yocto or in the component recipe to be able to take care of the change coming from yocto mainline.
